### PR TITLE
[LTC] Code-gen index_select

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This commit code-generated index_select.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestIndexSelect*

Fixes #65576.

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 1]
```diff
@@ -61,6 +61,26 @@ std::vector<c10::ScalarType> compute_dtype_embedding_dense_backward(const at::Te
...
 }
 
+std::vector<std::vector<int64_t>> compute_shape_index_select(const at::Tensor & self, int64_t dim,
+    const at::Tensor & index) {
+  // Based on definition of https://pytorch.org/docs/stable/generated/torch.index_select.html.
+  // Promote Rank 0 index tensor to a 1 * 1 tensor.
+  auto index_dim = index.dim() > 0 ? index.dim() : 1;
+  auto index_size = index.dim() > 0 ? index.size(0) : 1;
+  TORCH_CHECK(index_dim == 1);
+  TORCH_CHECK(dim >= 0 && dim < self.dim());
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:71]: isn't this check wrong?

I think we should support negative dim's.

```
>>> torch.index_select(x, 0, torch.tensor([1,2]))
tensor([[0., 1., 0., 0., 0.],
        [0., 0., 1., 0., 0.]])
>>> torch.index_select(x, 1, torch.tensor([1,2]))
tensor([[0., 0.],
        [1., 0.],
        [0., 1.],
        [0., 0.],
        [0., 0.]])
>>> torch.index_select(x, -1, torch.tensor([1,2]))
tensor([[0., 0.],
        [1., 0.],
        [0., 1.],
        [0., 0.],
        [0., 0.]])
```

See how I handled this in compute_shape_cat:

```
  int64_t rank = tensors[0].sizes().size();
  int64_t canonical_dim = dim % rank;
  if (canonical_dim < 0) {
      canonical_dim += rank;
  }
  ```

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:71]: you could make `canonical_dim` a helper func at the top of LazyShapeDtype.cpp, i think

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 2]
```diff
@@ -61,6 +61,26 @@ std::vector<c10::ScalarType> compute_dtype_embedding_dense_backward(const at::Te
   return {grad_output.scalar_type()};
 }
 
+std::vector<std::vector<int64_t>> compute_shape_index_select(const at::Tensor & self, int64_t dim,
+    const at::Tensor & index) {
+  // Based on definition of https://pytorch.org/docs/stable/generated/torch.index_select.html.
+  // Promote Rank 0 index tensor to a 1 * 1 tensor.
+  auto index_dim = index.dim() > 0 ? index.dim() : 1;
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:68]: when can index.dim() be 0?  I found that it is invalid to use this:

```
>>> torch.index_select(x, -1, torch.tensor())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: tensor() missing 1 required positional arguments: "data"
>>> torch.index_select(x, -1, torch.tensor([]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: index_select(): Expected dtype int32 or int64 for index
>>> torch.index_select(x, -1, torch.tensor([0]))
tensor([[1.],
        [0.],
        [0.],
        [0.],
        [0.]])
```

Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:68]: We have this test case `AtenLtcTsTensorTest.TestIndexSelectRank0` which seems passing all the time... Not sure...

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp (hunk 1 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:71]: Oh, negative dim. Let me adjust that.

**Review by wconstab — APPROVED** [review]

**Comment by alanwaketan** [comment]
Thanks @wconstab for approving this pull request.

File: lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp [hunk 3]
```diff
@@ -66,7 +66,7 @@ std::vector<std::vector<int64_t>> compute_shape_index_select(const at::Tensor &
     const at::Tensor & index) {
   // Based on definition of https://pytorch.org/docs/stable/generated/torch.index_select.html.
   // Promote Rank 0 index tensor to a 1 * 1 tensor.
-  dim = Helpers::GetCanonicalDimensionIndex(dim, self.dim());
+  dim = at::maybe_wrap_dim(dim, self);
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeDtype.cpp:69]: excellent!

**Comment by wconstab** [comment]
@alanwaketan did this pass CI? it looks like it is broken to me.  Seems like `  dim = at::maybe_wrap_dim(dim, self)` should probably be changed to call self.dim() or something?

**Comment by alanwaketan** [comment]
> @alanwaketan did this pass CI? it looks like it is broken to me. Seems like ` dim = at::maybe_wrap_dim(dim, self)` should probably be changed to call self.dim() or something?

I think it did: https://github.com/pytorch/pytorch/actions/runs/1410318375.

**Pull request merged**